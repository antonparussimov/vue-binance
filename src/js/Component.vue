<template>
  <transition :name="transition">
    <div tabindex="0"
         class="vld-overlay is-active"
         :class="{ 'is-full-page': isFullPage }"
         v-show="isActive"
         :aria-busy="isActive"
         aria-label="Loading"
         :style="{ zIndex }">
      <div class="vld-background"
           @click.prevent="cancel"
           :style="bgStyle">
      </div>
      <div class="vld-icon">
        <slot name="before"/>
        <slot name="default">
          <component :is="loader" :color="color" :width="width" :height="height"/>
        </slot>
        <slot name="after"/>
      </div>
    </div>
  </transition>
</template>

<script>
import {removeElement, HTMLElement} from './helpers.js'
import trapFocusMixin from './trapFocusMixin.js';
import Loaders from '../loaders/index.js';

export default {
  name: 'vue-loading',
  mixins: [trapFocusMixin],
  props: {
    active: Boolean,
    programmatic: Boolean,
    container: [Object, Function, HTMLElement],
    isFullPage: {
      type: Boolean,
      default: true
    },
    enforceFocus: {
      type: Boolean,
      default: true
    },
    lockScroll: {
      type: Boolean,
      default: false
    },
    transition: {
      type: String,
      default: 'fade'
    },
    /**
     * Allow user to hide the loader
     */
    canCancel: Boolean,
    /**
     * Do something on cancel
     */
    onCancel: {
      type: Function,
      default: () => {
      }
    },
    color: String,
    backgroundColor: String,
    blur: {
      type: String,
      default: '2px'
    },
    opacity: Number,
    width: Number,
    height: Number,
    zIndex: Number,
    loader: {
      type: String,
      default: 'spinner'
    }
  },
  data() {
    return {
      // Don't mutate the prop
      isActive: this.active
    }
  },
  components: Loaders,
  beforeMount() {
    // Insert the component in DOM when called programmatically
    if (this.programmatic) {
      if (this.container) {
        this.isFullPage = false;
        this.container.appendChild(this.$el)
      } else {
        document.body.appendChild(this.$el)
      }
    }
  },
  mounted() {
    // Activate immediately when called programmatically
    if (this.programmatic) {
      this.isActive = true;
    }

    document.addEventListener('keyup', this.keyPress)
  },
  methods: {
    /**
     * Proxy to hide() method.
     * Gets called by ESC button or when click outside
     */
    cancel() {
      if (!this.canCancel || !this.isActive) return;
      this.hide();
      this.onCancel.apply(null, arguments);
    },
    /**
     * Hide and destroy component if it's programmatic.
     */
    hide() {
      this.$emit('hide');
      this.$emit('update:active', false);

      // Timeout for the animation complete before destroying
      if (this.programmatic) {
        this.isActive = false;
        setTimeout(() => {
          this.$destroy();
          removeElement(this.$el)
        }, 150)
      }
    },
    disableScroll() {
      if (this.isFullPage && this.lockScroll) {
        document.body.classList.add('vld-shown');
      }
    },
    enableScroll() {
      if (this.isFullPage && this.lockScroll) {
        document.body.classList.remove('vld-shown');
      }
    },
    /**
     * Key press event to hide on ESC.
     *
     * @param event
     */
    keyPress(event) {
      // todo keyCode is deprecated
      if (event.keyCode === 27) this.cancel()
    },
  },
  watch: {
    active(value) {
      this.isActive = value
    },
    isActive(value) {
      if (value) {
        this.disableScroll();
      } else {
        this.enableScroll()
      }
    }
  },
  computed: {
    bgStyle() {
      return {
        background: this.backgroundColor,
        opacity: this.opacity,
        backdropFilter: `blur(${this.blur})`
      }
    }
  },
  beforeDestroy() {
    document.removeEventListener('keyup', this.keyPress);
  },
}
</script>
