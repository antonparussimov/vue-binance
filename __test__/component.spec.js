import {mount} from '@vue/test-utils';
import Component from '../src/js/Component';

describe('Loading component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(Component, {
      propsData: {
        active: true
      }
    });
  });

  test('renders when active', async () => {
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot()
  });

  test('renders when not active', async () => {
    wrapper.setProps({
      active: false
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot()
  });

});
