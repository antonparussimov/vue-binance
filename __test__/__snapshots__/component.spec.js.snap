// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component renders when active 1`] = `
<div
  aria-busy="true"
  aria-label="Loading"
  class="loading-overlay is-active is-full-page"
  name="fade"
  tabindex="0"
>
  <div
    class="loading-background"
  />
   
  <div
    class="loading-icon"
  >
    <svg
      height="64"
      stroke="#666"
      viewBox="0 0 38 38"
      width="64"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <g
          stroke-width="2"
          transform="translate(1 1)"
        >
          <circle
            cx="18"
            cy="18"
            r="18"
            stroke-opacity=".25"
          />
           
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
          >
            <animatetransform
              attributeName="transform"
              dur="0.8s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
        </g>
      </g>
    </svg>
  </div>
</div>
`;

exports[`Loading component renders when not active 1`] = `
<div
  aria-label="Loading"
  class="loading-overlay is-active is-full-page"
  name="fade"
  style="display: none;"
  tabindex="0"
>
  <div
    class="loading-background"
  />
   
  <div
    class="loading-icon"
  >
    <svg
      height="64"
      stroke="#666"
      viewBox="0 0 38 38"
      width="64"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <g
          stroke-width="2"
          transform="translate(1 1)"
        >
          <circle
            cx="18"
            cy="18"
            r="18"
            stroke-opacity=".25"
          />
           
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
          >
            <animatetransform
              attributeName="transform"
              dur="0.8s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
